api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,546 @@ package v1alpha1
-// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=externaldnses,scope=Cluster
+// +kubebuilder:subresource:status
+//
+// ExternalDNS describes a managed ExternalDNS controller instance for a cluster.
+// The controller is responsible for creating external DNS records in supported
+// DNS providers based off of instances of select Kubernetes resources.
+type ExternalDNS struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the ExternalDNS.
+	Spec ExternalDNSSpec `json:"spec"`
+	// status is the most recently observed status of the ExternalDNS.
+	Status ExternalDNSStatus `json:"status"`
+}
 
-// ExternalDNSSpec defines the desired state of ExternalDNS
+// ExternalDNSSpec defines the desired state of the ExternalDNS.
 type ExternalDNSSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
+	// Domains specifies which domains that ExternalDNS should
+	// create DNS records for. Multiple domain values
+	// can be specified such that subdomains of an included domain
+	// can effectively be ignored using the "Include" and "Exclude"
+	// domain filter options.
+	//
+	// An empty list of domains means ExternalDNS will create
+	// DNS records for any included source resource regardless
+	// of the resource's desired hostname.
+	//
+	// Populating Domains with only excluded options means ExternalDNS
+	// will create DNS records for any included source resource that do not
+	// match the provided excluded domain options.
+	//
+	// Excluding DNS records that were previous included via a resource update
+	// will *not* result in the original DNS records being deleted.
+	//
+	// +optional
+	Domains []ExternalDNSDomain `json:"domains,omitempty"`
+
+	// Provider refers to the DNS provider that ExternalDNS
+	// should publish records to. Note that each ExternalDNS
+	// is tied to a single provider.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Provider ExternalDNSProvider `json:"provider"`
+
+	// Source describes which source resource
+	// ExternalDNS will be configured to create
+	// DNS records for.
+	//
+	// Multiple ExternalDNS CRs must be
+	// created if multiple ExternalDNS source resources
+	// are desired.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Source ExternalDNSSource `json:"source"`
 
-	// Foo is an example field of ExternalDNS. Edit externaldns_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// Zones describes which DNS Zone IDs
+	// ExternalDNS should publish records to.
+	//
+	// Updating this field after creation
+	// will cause all DNS records in the previous
+	// zone(s) to be left behind.
```

Comment:
The orphan policy on everything concerns me in terms of trying this feature out and later removing it.  Having an orphan policy makes sense, but I would expect a cleanup policy to be the default.